src/FuseGPUThreadLoops.cpp

```diff
@@ -262,19 +266,54 @@ class ReplaceForWithIf : public IRMutator {
 class ExtractSharedAllocations : public IRMutator {
     using IRMutator::visit;
 
+    struct IntInterval {
+        IntInterval() : IntInterval(0, 0) {}
+        IntInterval(int min, int max) : min(min), max(max) {}
+        int min;
+        int max;
+    };
+
     struct SharedAllocation {
         string name;
         Type type;
         Expr size;
+        IntInterval liveness; // Start and end of the barrier stage at which this allocation is used.
+    };
+
+    struct AllocGroups {
```

Comment:
Should this be named `AllocGroup`? `AllocGroups` plural makes me think there is a set of groups, but I think this struct just represents one group of allocations.